### PR TITLE
HPCC-28911 Allow version to be overridden but keep default repo

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: ecl/eclcc/eclcc.cpp [hunk 1]
```diff
@@ -3090,11 +3097,24 @@ int EclCC::parseCommandLineOptions(int argc, const char* argv[])
...
         if (!strchr(optQueryMainPackage.str(), '#'))
-            optQueryMainPackage.append("#").append(optQueryMainPackageVersion);
+        {
+            if (!optQueryMainPackageVersion.isEmpty())
+                optQueryMainPackage.append("#").append(optQueryMainPackageVersion);
+        }
+    }
+    else
+    {
+        if (!optDefaultRepo.isEmpty())
```
Comment (shamser) [inline ecl/eclcc/eclcc.cpp:3116]: Not sure this is this needed given that optQueryMainPackage is being set to optDefaultRepo **in line 3101**?

File: ecl/eclcc/eclcc.cpp [hunk 2]
```diff
@@ -2862,12 +2863,18 @@ int EclCC::parseCommandLineOptions(int argc, const char* argv[])
...
             const char * arg = tempArg;
             const char * at = strchr(arg, '@');
+            const char * hash = strchr(arg, '#');
             if (at)
             {
                 optQueryMainAttribute.set(arg, at - arg);
                 optQueryMainPackage.set(at+1);
                 optExplicitMainPackage = true;
             }
+            else if (hash)
```
Comment (shamser) [inline ecl/eclcc/eclcc.cpp:2873]: What should happen if they provide both @ and #?

**Review by shamser — CHANGES_REQUESTED** [review]

File: ecl/eclcc/eclcc.cpp (hunk 2 above)
Comment (ghalliday) [reply ecl/eclcc/eclcc.cpp:2873]: A # can already follow an @ e.g., x@ghalliday/ecl-demo-d#version3

**Comment by ghalliday** [comment]
@shamser please re-review.  Main logic has been rewritten and clarified.  Additional tests added.

File: ecl/regress/testrepo.sh [hunk 3]
```diff
@@ -0,0 +1,98 @@
+# Script file to testing different variants of sumbitting queries
```
Comment (ghalliday) [inline ecl/regress/testrepo.sh:1]: typo: submitting.  Will fix when I squash,

File: ecl/eclcc/eclcc.cpp [hunk 4]
```diff
@@ -3097,30 +3097,60 @@ int EclCC::parseCommandLineOptions(int argc, const char* argv[])
...
-        optQueryMainPackageVersion.set(optDefaultRepoVersion);
+    {
+        StringBuffer temp;
+        const char * repo = optDefaultRepo.str();
+        const char * hash = strchr(repo, '#');
+        if (hash)
+        {
+            if (!optDefaultRepoVersion.isEmpty())
+            {
+                //Unusual situtations, but resolvie it consistently
```
Comment (shamser) [inline ecl/eclcc/eclcc.cpp:3111]: typos: situtations & situtations

File: ecl/eclcc/eclcc.cpp [hunk 5]
```diff
@@ -3097,30 +3097,60 @@ int EclCC::parseCommandLineOptions(int argc, const char* argv[])
...
+        const char * repo = optQueryMainPackage.str();
+        const char * hash = strchr(repo, '#');
+        const char * newVersion = nullptr;
+        if (hash)
         {
+            //only thing that takes precedence is optQueryMainPackageVersion
             if (!optQueryMainPackageVersion.isEmpty())
-                optQueryMainPackage.append("#").append(optQueryMainPackageVersion);
+            {
+                optQueryMainPackage.setLength(hash-repo); // Remove ther version - ready for replacing below.
```
Comment (shamser) [inline ecl/eclcc/eclcc.cpp:3143]: typo: ther

File: ecl/regress/demomain.ecl [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import format;
```
Comment (shamser) [inline ecl/regress/demomain.ecl:1]: is this ecl related to this PR?

**Review by shamser — APPROVED** [review]
There's couple of other typos.  It could probably be done also when squashing.
